### ISSUE TITLE
`machinctl shell` doesn't work in containers that aren't booted

### ISSUE THREAD

**Issue by larskarlitski** [systemd/systemd#12785]
**systemd version the issue has been seen with**
241

**Used distribution**
 Fedora

I've spawned a container from a directory (with a fedora tree):
```
# systemd-nspawn --directory tree --machine foo
Spawning container foo on /home/lars/tree.
Press ^] three times within 1s to kill container.
[root@foo ~]# 
```

I expect to be able to enter the container with `machinectl shell`, but it fails:
```
# sudo machinectl shell foo
Failed to get shell PTY: Input/output error
```
Entering with `nsenter` works:
```
# nsenter -a --wd=$PWD/tree -t `machinectl show foo -p Leader --value`
[root@foo /]# 
```

**Comment by refi64** [comment]
This is probably [this SELinux issue](https://bugzilla.redhat.com/show_bug.cgi?id=1257990), which has been fixed in the rawhide policies, but I don't think it's made it into stable yet. Try setting it to permissive to verify.

**Comment by larskarlitski** [comment]
> This is probably [this SELinux issue](https://bugzilla.redhat.com/show_bug.cgi?id=1257990), which has been fixed in the rawhide policies, but I don't think it's made it into stable yet. Try setting it to permissive to verify.

I'm running Fedora 30, which has the fix for that issue. I've tried anyway and I'm getting the same error with SELinux in permissive mode. I think `machinectl` is trying to access D-Bus inside the container when spawning bash in a transient unit.

**Comment by poettering** [comment]
Fixing that is nasty. nsenter is a hack since it inserts a foreign, long lasting process into the namespace, that inherits only a very very limited set of properties from the container (i.e. the ones the tool explicitly changes), but almost everything else comes from the host. i.e. all the many many props, such as scheduling, cgroup stuff, security labels, audit, yadda yadda come from the host, and doing tht for long running stuff like that is really, really messy. Moreover, processes like that have PID 0 as parent, and not PID 1 of the container, which is very weird. In addition the process in the container (i.e. bash given that you didn't boot it up properly) doesn't really expect to gain a new child just like that if you double fork). "machinectl shell" hence works differently: i talks to PID 1 in the container, and asks it to fork something off, while providing stdin/stdout/stderr fds. that way, the processes run that way are proper members of the containers, and everything comes from the container, except the bare minimum, the connection fds...

Hence, in my eyes, "nsenter" (and anything working like it) are mostly just debugging tools, not a clean operation. But of course this means that "machinectl shell" only works for containers with PID 1 running properly, with an IPC interface one can talk to (which means systemd) that can start stuff for us on request. 

I think it would be great if we could fix this properly. For that, I figure some kernel work would be needed. Specifically, we'd probably need a new system call: fforkandexec() or so, that combines a fork() and an exec() while allowing specification of two things: the pid to use as parent and to inherit all props from, and the fd of the binary to execute. That would allow us to safely, and correctly insert a process into a container, without ending up with a disastrous mess of some properties from the outside and some of the inside...

**Comment by larskarlitski** [comment]
Oh interesting. Thanks for the context. I probably want to pass `--register=no` then. Could that be the default for containers that aren't booted?

I do think there could be a better error message here :) Otherwise, feel free to close this bug.

**Comment by poettering** [comment]
Yeah, I figure it would make a ton of sense to at least output a clear message explaining the situation and maybe even with a helper pre-packaged nsenter command line in it, that comes with all the right parameters.

**Comment by asb** [comment]
As an alternative approach that doesn't require kernel changes, might it make sense to look at expanding nspawn-stub-pid1 so that it can support spawning new shells? I understand there's no desire to expand it into a full init, but it feels like there's a chance a small amount of extra functionality would improve experience for cases like this, or when dealing with non-systemd containers.

**Comment by keszybz** [comment]
> expand nspawn-stub-pid1 so that it can support spawning new shells

Yep, that'd be very nice.

> maybe even with a helper pre-packaged nsenter command line in it

Well, if we are smart enough to figure out the nsenter command, then we should just do it. No need to "educate" the user.

So I'd like to see the following behaviour:
```
$ systemd-nspawn -M foo --boot
$ machinectl shell foo
(foo prompt)
$ systemd-nspawn -M foo --as-pid2
$ machinectl shell foo
(foo prompt)
$ systemd-nspawn -M foo
$ machinectl shell foo
Warning: machine foo does not support ... (some proper error message here).
               A shell will spawned using from the outside, but it is a foreign process that inherits the environment from the outside.
(foo prompt)
```

I'll take this off the v244 milestone, since it's an RFE not a bug. Pull requests are welcome ;)

**Comment by keithy** [comment]
Not finding a way to enter portable services that have notbooted/failed either. Ideas?

**Comment by boucman** [comment]
I am not sure "entering" a portable service makes sense.. a portable service is a single application, totally self-contained. It is very likely there is no shell in there.

moreover, you don't boot a portable service. it doesn't have a init at all.
It would be nice if there was an easy, systemd-provided  way to enter the proper mount namespace (without the chroot) in order to explire the content, though...

**Comment by keithy** [comment]
So the option of putting a shell in as a portable service operating within that namespace, would allow you to see things from that point of view.
